lib/metadata/DataType.hpp

```diff
@@ -0,0 +1,198 @@
+#ifndef DATATYPE_HPP
+#define DATATYPE_HPP
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <type_traits>
+
+// Define a macro `KOTEKAN_FLOAT16` specifying whether we support a float16 type.
+// Define a type `float16_t` if we support it.
+
+#if defined WITH_CUDA
+// If we use CUDA, use its float16 type
+#include <cuda_fp16.h>
+using float16_t = __half;
+#define KOTEKAN_FLOAT16 1
+#else
+// If we don't use CUDA, see whether the compiler supports it
+#include <float.h>
+#if defined __FLT16_MAX__
+using float16_t = _Float16;
+#define KOTEKAN_FLOAT16 1
+#else
+// There is no float16 type
+#define KOTEKAN_FLOAT16 0
+#endif
+#endif
+
+namespace chord {
```

Comment:
I'm not sure it makes sense to use namespace `chord` here, in theory these functions will get used by HIRAX and CHIME at some point.

Reply:
True! I should have put this into `kotekan` instead.

However, we should use a namespace, just to avoid conflicts with other libraries.